### ISSUE TITLE
Guile

### ISSUE THREAD

**Issue by mkohlhaas** [void-linux/void-packages#23056]
We should provide packages for Guile 3.0.x series and 2.2.x series. The current Guile package is based on version 2.0.14 and needs updating.

Maintainer seems to be orphaned. But I'm not sure about it. If yes I could help out.

**Comment by mkohlhaas** [comment]
Actually, we should only provide 3.0.x series as this is the stable release series. Though bugfix versions for 2.2.x series are still provided.

So best option is just to switch to Guile 3.0.x series.

**Comment by ericonr** [comment]
Updating to 2.2.x has been very hard, as seen in https://github.com/void-linux/void-packages/pull/22304#issuecomment-633793722 . If you'd like to try it, that would be awesome.

I have #22359 going on, but gnutls and Guix have been a bit of a pain.

**Comment by mkohlhaas** [comment]
Hi Érico,

I'd go straight for the officially stable 3.0.x release of you don't mind. Or are there any reasons against it?

**Comment by ericonr** [comment]
None that I know of. That's the one I'm using in my PR, isn't it?

I'm not sure we want guile3 though, unless other applications are using it. aisleriot could be switched to it, but that was the only one. There's Guix too, but I need to fix cross compilation, otherwise it will basically be a package for x86_64 and aarch64 (and ppc*), which I don't like much.

**Comment by mkohlhaas** [comment]
Hi Érico,

I was just checking the packages using Guile. Most of them either support Guile version 2.2 and/or 3.0 which makes perfectly sense as they are supported and versions 2.0 and 1.8 are obsolete. The last 2.0 release has been made in 2017, the last 1.8 release in 2010.

A software package supporting both Guile versions 3.0 and 2.2 should use 3.0 as this is refered to by the Guile project as the *latest stable version*. (Version 2.2 is the *legacy version*.) Some of the existing packages using Guile should then be updated too - if possible - in order to reference to these two latest Guile versions.

Currently only our `Lilipond` package uses Guile version 1.8.

I couldn't find package `Guix`. But I guess you are refering to [GNU Linux](https://guix.gnu.org/)?

These are the packages using Guile with their highest supported Guile version checked with the actual sources:
Package|Highest supported Guile version|Status
-----------| ----------------------------------------|--------
`aisleriot`|3.0|OK
`autogen`|2.2|OK
`cross-arm-none-eabi-gdb`|Package comment: "don't enable guile and python interfaces until they are moved into platform independent packages"|↯ (GNU GDB needs [this patch](https://sourceware.org/pipermail/gdb-patches/2020-June/169434.html) to support Guile 3.0!)
`geda`|2.0|↯ ([Geda Bug Report](https://bugs.launchpad.net/geda/+bug/1782306))
`gnucash`|2.2|OK (GnuCash README: Guile compiled with regex support enabled)
`gnutls`|2.2|OK
`guile-wisp`|2.2|OK
`lilypond`|2.2|OK
`make`|3.0|OK
`trackballs`|2.2|OK
`xbindkeys`|3.0|OK

Status OK means package either supports Guile version 3.0 or 2.2 (or both).

**Comment by flexibeast** [comment]
> I couldn't find package Guix. But I guess you are refering to GNU Linux?

In this case, @ericonr is referring to [the Guix package manager](https://en.wikipedia.org/wiki/GNU_Guix) used by the Guix distro; see @ericonr's PR https://github.com/void-linux/void-packages/pull/22359.

**Comment by mkohlhaas** [comment]
So he wants the Guix package manager on top of Void Linux?

**Comment by mkohlhaas** [comment]
Refering to [New package Guile 3](https://github.com/void-linux/void-packages/pull/22304) and [Updating from 2.0 to 2.2](https://github.com/void-linux/void-packages/pull/6524) (which is not possible):

We need separate Guile packages:
Guile Packe|Void Package Needed|Reason|Supported
---------------|----------------------------|----------|-------------
3.0|Yes|Officially stable release|Yes
2.2|Yes|Legacy release (still used by many packages)|Yes
2.0|Yes|Obsolete but unfortunately still used by eg. Geda and GNU GDB (at least without [this patch](https://sourceware.org/pipermail/gdb-patches/2020-June/169434.html) from **last week**!)|No
1.8|No|Obsolete and there are no packages needing 1.8|No

Guile 3.0 should be the default. That means when I type `guile` in the command line I should end up in Guile 3.0:
Guile Package|Binary Names
------------------|------------------
3.0|guile<br>guile3<br>guile3.0
2.2|guile2<br>guile2.2
2.0|guile2.0
1.8|guile1.8

**In summary: we need separate Guile packages for versions 2.0, 2.2 and 3.0!**
(Of course we can keep package `guile1.8` as it won't do any harm.)

**Comment by newbluemoon** [comment]
Where did you find lilypond works with guile > 1.8? The `INSTALL.txt` in the latest tarball says:

> • Guile (http://www.gnu.org/software/guile/guile.html) Use version
     1.8.8.  Version 2.x of Guile is not currently supported.

However, there is a configure option to build lilypond with guile 2.x. Last time I tested it the build succeeded, but the result was flawed and ugly in no comparison to using 1.8. So lilypond still depends on guile 1.8.

**Comment by ericonr** [comment]
> gnutls | 2.2

GnuTLS builds fine with Guile 3.0.x, and worked fine for Guix. The only issue I can't get past with it is that of cross compilation.

**Comment by ericonr** [comment]
Regarding the situation as a whole, I don't know what the policy for changing this stuff is. Alternatives for guile 2 and 3 might work well: have `guile3` and `guile2` executables, then make a symlink to `guile`.

So basically we need 2 non updating packages + a legacy release + stable release? Not sure if this maintenance burden will be accepeted.

**Comment by mkohlhaas** [comment]
@newbluemoon 
Was using the tarball from the [releases page](https://github.com/lilypond/lilypond/releases) on Github which does not have a `INSTALL.txt` file and not from [Lilypong download page](http://lilypond.org/downloads/sources/v2.21/) . Only checked the configuration options from `aclocal.m4`. Also checking with their bug database it seems they having problems updating to Guile versions 2.0 and 2.2. I have updated the table.

**Comment by mkohlhaas** [comment]
> > gnutls | 2.2
> 
> GnuTLS builds fine with Guile 3.0.x, and worked fine for Guix. The only issue I can't get past with it is that of cross compilation.

Accepted. Gnutls is acutally searching for Guile version 3.0. Updated the list.

**Comment by mkohlhaas** [comment]
> So basically we need 2 non updating packages + a legacy release + stable release? Not sure if this maintenance burden will be accepeted.

Python has a similar issue. (Their default version is 2.x and not 3.x.)

**Comment by mkohlhaas** [comment]
> > I couldn't find package Guix. But I guess you are refering to GNU Linux?
> 
> In this case, @ericonr is referring to [the Guix package manager](https://en.wikipedia.org/wiki/GNU_Guix) used by the Guix distro; see @ericonr's PR #22359.

I like this Guix thingy. Brilliant idea! Thanks for pointing me to it!

**Comment by ericonr** [comment]
> Python has a similar issue. (Their default version is 2.x and not 3.x.)

That's still four packages, Python only needs two. And we have a lot of dependants on Python too, which isn't the case for Guile.

Still, if you want to package them , you're welcome to. I've done some of the work for Guile 3, up to having a working Guix package for both x86_64 and aarch64.  What I'm currently missing is GnuTLS compiling for 32bit architectures.

**Comment by mkohlhaas** [comment]
Was just getting my feet wet by installing Guile. The point is Guile doesn't do very well with different co-existing full Guile versions to say the least. It's basically impossible. Even the current situation with our different packages for different Guile versions is far from ideal.

So what one actually needs is one main Guile 3.0 package and the other versions as lib packages for runtime support. Also Guile devel packages that will be used as hostmakedpends 3rd party packages like eg. `Lilypond` that use Guile as an extension language.

That means as an end user you should be able to install Guile which will install Guile 3.0. End users would be allowed to install the libs for other Guile versions namely 1.8, 2.0 and 2.2. So they can run  software that depend on it eg. `Lilypond` but are not the allowed to install whole Guile packages of these versions with interactive Guile interpreters. 

In summary I guess we should have following packages:
Guile version|main|lib|devel
--------------------|--------|---|-----
3.0|✓|-|-
2.2|empty|✓|✓
2.0|empty|✓|✓
1.8|empty|✓|✓

Guile version|main|lib|devel
--------------------|--------|---|-----
3.0|everything|-|-
2.2|-|libraries|headers, pkg-config, ac files, support binaries
2.0|-|libraries|headers, pkg-config, ac files, support binaries
1.8|-|libraries|headers, pkg-config, ac files, support binaries

Support binaries are things like `guile-config`, `guile-snarf`.

I didn't check so far whether the `lib` packages overlap.

**Comment by ericonr** [comment]
Have you looked into the PR I made? I'm packaging Guile 3 stuff with a `3` suffix for all binaries, that way they don't conflict with the current version. Adding an alternatives package could be a way to allow for a default package.

**Comment by mkohlhaas** [comment]
Yes, sure. Just changing the default binaries with `--program-suffix` is not enough. We have that already in the Guile 1.8 package. The pkg-config files are not correct. And we need at least `xbps-alternative` to distinguish binaries for the end user.

I am afraid there is no shortcut. This is what is necessary:

Guile version|main|lib|devel
--------------------|--------|---|-----
3.0|✓|✓|✓|
2.2|✓|✓|✓|
2.0|✓|✓|✓|
1.8|✓|✓|✓|

Maybe even a `doc` subpackage but mainly for cosmetics reasons.
